### PR TITLE
Migration of "Design -> pages -> add new / edit page category"

### PR DESCRIPTION
**Pull request by tomas862**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrates cms page category add and edit forms only.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10567
| How to test?  | Test only add and edit forms. Should work exactly the same

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12647)

### CODE REVIEW

**Comment by tomas862** [comment]
@matks  ready for review as well ;)!

File: src/Adapter/CMS/PageCategory/CategoriesProvider.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * @param int $contextLanguageId
+     * @param array $contextShopIds
+     */
+    public function __construct(
+        $contextLanguageId,
+        array $contextShopIds
+    ) {
+        $this->contextLanguageId = (int) $contextLanguageId;
+        $this->contextShopIds = array_map(function ($item) { return (int) $item; }, $contextShopIds);
```
Comment (matks) [inline src/Adapter/CMS/PageCategory/CategoriesProvider.php:58]: 😭

File: src/Core/Domain/CmsPageCategory/DTO/EditableCmsPageCategory.php [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+     * @var array
+     */
+    private $localisedMetaKeywords;
+
+    /**
+     * @var array
+     */
+    private $localisedFriendlyUrl;
+    /**
+     * @var array
```
Comment (matks) [inline src/Core/Domain/CmsPageCategory/DTO/EditableCmsPageCategory.php:72]: The fact that it's an array was very strange to me 😅 I guess translation makes things complex

Comment (matks) [reply src/Core/Domain/CmsPageCategory/DTO/EditableCmsPageCategory.php:72]: We're gonna have to define a convention for this usecase. So far although this is an array, we keel a singular variable name. I dont know if it's relevant, but whatever we choose, we need to write it down.

File: src/Core/Domain/CmsPageCategory/Exception/CmsPageCategoryConstraintException.php [hunk 3]
```diff
@@ -31,5 +31,54 @@
...
+
+    /**
+     * @var int is used to validate meta title for specific regex pattern
+     */
+    const INVALID_META_TITLE = 7;
+
+    /**
+     * @var int Is used to validate meta description for specific regex pattern
+     */
+    const INVALID_META_DESCRIPTION = 8;
```
Comment (matks) [inline src/Core/Domain/CmsPageCategory/Exception/CmsPageCategoryConstraintException.php:73]: 👍

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php [hunk 4]
```diff
@@ -98,20 +100,39 @@ public function indexAction(CmsPageCategoryFilters $filters, Request $request)
...
+                    $this->trans('Successful creation.', 'Admin.Notifications.Success')
+                );
+
+                return $this->redirectToIndexPageById($result->getIdentifiableObjectId());
+            }
+        } catch (CmsPageCategoryException $exception) {
+            $this->addFlash('error', $this->handleException($exception));
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/Design/Cms/create_category.html.twig', [
```
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:133]: 👍

File: tests/Unit/Core/Domain/CmsPageCategory/Command/EditCmsPageCategoryCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace Tests\Unit\Core\Domain\CmsPageCategory\Command;
+
+use PHPUnit\Framework\TestCase;
+
+class EditCmsPageCategoryCommandTest extends TestCase
```
Comment (matks) [inline tests/Unit/Core/Domain/CmsPageCategory/Command/EditCmsPageCategoryCommandTest.php:34]: Looks like a nice beginning, why sop there 😄 ?

File: src/Adapter/CMS/PageCategory/CategoriesProvider.php (hunk 1 above)
Comment (tomas862) [reply src/Adapter/CMS/PageCategory/CategoriesProvider.php:58]: You know I need to casts things before using them in sql - but be perfect to receive context language id or shop ids in already type safe form but unfortunately we can't do that right now or can we ? :thinking:

File: tests/Unit/Core/Domain/CmsPageCategory/Command/EditCmsPageCategoryCommandTest.php (hunk 5 above)
Comment (tomas862) [reply tests/Unit/Core/Domain/CmsPageCategory/Command/EditCmsPageCategoryCommandTest.php:34]: forgot to continue - added some excepion checking tests for both `EditCmsPageCategoryCommand` and `AddCmsPageCategoryCommand` :+1:

**Comment by tomas862** [comment]
@matks done - added missing tests :)

File: src/Adapter/CMS/PageCategory/CategoriesProvider.php (hunk 1 above)
Comment (matks) [reply src/Adapter/CMS/PageCategory/CategoriesProvider.php:58]: Yes I know. I sob but I accept (that's why its a comment, not a requested change 😭 ) it.

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
@LouiseBonnard You can perform a wording review, this PR code is validated ✅

**Comment by marionf** [comment]
@tomas862 

1. When I click on "Help", I have this 

capture d'écran_1185

It should opened like below instead

capture d'écran_1186

2. The string and the language are not displayed

capture d'écran_1187

Can we display the string with the error and the language concerned like below ?

capture d'écran_1189

3. For the meta keywords field, can we have it like on other pages ?

capture d'écran_1188

**Comment by tomas862** [comment]
@marionf  done

1. Can't re-create it - maybe its cache for you or assets are not recompiled. Anyway, please run the command `npm run build` in order to apply new changes
2. Fixed it ;)
3. Done

**Comment by marionf** [comment]
All good, thank you @tomas862 :+1:

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig [hunk 6]
```diff
@@ -163,13 +163,16 @@
           {% endset %}
       {% endif %}
 
+      <template class="js-popover-error-content" data-id="{{ form.vars.id }}">
```
Comment (matks) [inline src/PrestaShopBundle/Resources/views/Admin/macros.html.twig:166]: I did not know `template` keyword, what is its behavior ? I cant find it in https://twig.symfony.com/doc/

File: admin-dev/themes/new-theme/js/pages/cms-page/index.js [hunk 7]
```diff
@@ -65,4 +66,11 @@ $(() => {
   shopChoiceTree.enableAutoCheckChildren();
 
   new TranslatableInput();
+
+  new TaggableField({
```
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/cms-page/index.js:70]: This is so cool that we can reuse our components so easily 😊 it means the job was well done

👏

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig (hunk 6 above)
Comment (tomas862) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig:166]: its perfect for holding the content which only is being used in javascript. https://www.w3schools.com/tags/tag_template.asp . Its same as div but I think its purpose is to not display the content for the client untill it is needed :)

File: admin-dev/themes/new-theme/js/pages/cms-page/index.js (hunk 7 above)
Comment (tomas862) [reply admin-dev/themes/new-theme/js/pages/cms-page/index.js:70]: yeah indeed , took me whole minute to add taggable field :D

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Thank you @tomas862 !

**Comment by tomas862** [comment]
Thank you @marionf  and @matks !

**Pull request merged**